Add a case in Remote Execution Workflow (#1646)
theforeman/foreman-documentation: guides/common/modules/con_remote-execution-workflow.adoc

[id="remote-execution-workflow_{context}"]
= Remote Execution Workflow

> {Project} prefers remote execution on host's registered_through proxy.

Author: > what is "registered_through"? Please reword to make it less technical.

The original idea from a reporter was to keep it like this. But, we are Okay to reword it.

> describe what REX is

We have used the phrase **Remote Execution** instead of REX.

> how SmartProxies play their part to run jobs on managed hosts.

Will again reword it. :+1:

> You must *enable remote_execution_prefer_registered_through_proxy* manually from *Administer > Settings > Content*.

Author: Addressed :+1:

> This will allow selecting a {SmartProxies} to perform the REX job on the hosts to which they are registered to.

Reviewer: ```suggestion
This will select the {SmartProxy} to which the hosts are registered, to perform the REX job.
```

Reviewer: @mjivraja I think this sounds better. What do you think @maximiliankolb ?

Author: Thanks @AkshayGadhaveRH 
Wait for Maxi if he has something for us.

Reviewer 2: I think the whole first paragraph should be reworded.

a) describe what REX is
b) how SmartProxies play their part to run jobs on managed hosts.

Assuming a host is always only registered to one smartproxy, either Foreman or a Smart Proxy, then describing that you _can_ use the "registered through" Smart Proxy sounds weird to me.

Author: Addressed. :+1: 
Kept `SmartProxies`

> For the case where the above setting is disabled, you can run a remote job on hosts, for every host, {Project} performs the following actions to find a remote execution {SmartProxy} to use.

Reviewer: ```suggestion
For the case where the above setting is disabled, for every host, {Project} performs the following actions to find a remote execution {SmartProxy} to use.
```

Author: Addressed :+1:

[id="remote-execution-workflow_{context}"]
= Remote Execution Workflow

> {Project} prefers remote execution on hosts that are registered through proxy.

Reviewer 2: ```suggestion
{Project} prefers remote execution on hosts that are registered through a {SmartProxy}.
```

IMHO this is not true, because `Prefer registered through proxy for remote execution` defaults to "no" on Foreman 3.3.

Author: Shall I go with the `Smart Proxy` that you suggested?

Reviewer 2: I still believe this is not entirely true because you want to have this change cherry-picked to 3.3 & I've checked the default behaviour on Foreman.

Author: Reworded that line :+1:

> Remote execution can be performed on the hosts that are registered through a {SmartProxy}.

Reviewer 2: You can also perform REX directly through `{ProjectServer}`.

Reviewer 3: I'm not 100% sure about what those curly placeholders get filled with, but REX jobs always go through a smart proxy, even though it may be colocated on the same machine

Author: > I'm not 100% sure about what those curly placeholders get filled with.

Curly brackets will replace the Project name as those are attributes. :slightly_smiling_face: 

> but REX jobs always go through a smart proxy, even though it may be colocated on the same machine

@maximiliankolb I am confused :smile: +

Reviewer 2: I guess what I'm trying to say: You can also attach a host to Foreman; so REX works through Foreman's internal Smart Proxy.

Foreman would be your ProjectServer, but not a "SmartProxyServer".

```suggestion
Remote execution can be performed on managed hosts through {SmartProxies}.
```

Author: Addressed :+1:

> You must enable *remote_execution_prefer_registered_through_proxy* manually.

Reviewer 2: This still needs to be reworded IMHO.

When describing in this concept _what_ REX is, we should not immediately start with "you'll probably what to use this very special feature; here's how to activate it".

Can you please provide some more details on _why_ we recommend enabling `remote_execution_prefer_registered_through_proxy`?

I'd also switch the order and add this below the procedure starting in line 11.

Author: I guess we dropped the idea of `recommending`. Instead, we kept that way optional for end-user. I hope I understood correctly from earlier comments. 

For other changes that you suggested, I am Okay with that :+1:

Author: Hi @maximiliankolb 

After my further discussion with SME, we are in agreement to provide some more details on why we recommend `enabling remote_execution_prefer_registered_through_proxy`?

I will add more text for the same. :+1:

Author: Addressed :+1:

[id="remote-execution-workflow_{context}"]
= Remote Execution Workflow

Remote execution can be performed on managed hosts through {Project} as well as {SmartProxies}.
Clients registered through one {SmartProxy} might not be able to communicate through other {SmartProxies} or the {project} directly and vice-versa.
As a result, the remote execution tasks must be performed through respective {SmartProxies} only that can communicate with the client host.
Therefore, enabling _remote_execution_prefer_registered_through_proxy_ should allow {Project} to run REX jobs on clients that are registered through their respective {SmartProxies}.

> You must enable *remote_execution_prefer_registered_through_proxy manually*.

Author: Addressed :+1:

> Navigate to *Administer > Settings*, enable the option from *Content* tab.

Reviewer 2: ```suggestion
Navigate to *Administer > Settings* and on the *Content* tab, set `remote_execution_prefer_registered_through_proxy` to true.
```

Author: Addressed :+1:

. {Project} finds remote execution {SmartProxies} assigned to these subnets.
. From this set of {SmartProxies}, {Project} selects the {SmartProxy} that has the least number of running jobs.
By doing this, {Project} ensures that the jobs load is balanced between remote execution {SmartProxies}.

> If you have enabled *remote_execution_prefer_registered_through_proxy*,
> {Project} runs the REX job using the {SmartProxy} the host is registered to.

Reviewer 2: ```suggestion
If you have enabled *remote_execution_prefer_registered_through_proxy*, {Project} runs the REX job using the {SmartProxy} the host is registered to.
```

Author: Thanks :+1:

. {Project} finds remote execution {SmartProxies} assigned to these subnets.
. From this set of {SmartProxies}, {Project} selects the {SmartProxy} that has the least number of running jobs.
By doing this, {Project} ensures that the jobs load is balanced between remote execution {SmartProxies}.

> 
> However, there is one more way to run REX jobs apart from the default one mentioned above.
> Remote execution can be performed on managed hosts through {Project} as well as {SmartProxies}.
> Clients registered through one {SmartProxy} might not be able to communicate through other {SmartProxies} or the {project} directly and vice-versa.
> As a result, the remote execution tasks must be performed through respective {SmartProxies} only that can communicate with the client host.
> Therefore, enabling _remote_execution_prefer_registered_through_proxy_ should allow {Project} to run REX jobs on clients that are registered through their respective {SmartProxies}.

Reviewer 2: ```suggestion
If you have enabled _Prefer registered through proxy for remote execution_, {Project} runs the REX job using the {SmartProxy} the host is registed to.
```

> For this option, you must enable *remote_execution_prefer_registered_through_proxy* manually.
> Navigate to *Administer > Settings* and on the *Content* tab, set `remote_execution_prefer_registered_through_proxy` to true.
> This will select the {SmartProxies} to which the hosts are registered, to perform the REX job.

Reviewer 2: ```suggestion
By default, _Prefer registered through proxy for remote execution_ is set to *No*.
To enable it, in the {ProjectWebUI}, navigate to *Administer* > *Settings*, and on the *Content* tab, set `remote_execution_prefer_registered_through_proxy` to *Yes*.
This ensures that {Project} performs REX jobs on hosts by the {SmartProxy} to which they are registered to.
```

> Therefore, you can use either way to perform REX jobs on hosts.
> 

Reviewer 2: ```suggestion
```
